### DISCUSSION TITLE
loading multiple, separate instances of an BPF_MAP_TYPE_XSKMAP?

### DISCUSSION THREAD

**Discussion by thediveo** [cilium/ebpf#1481] [Q&A] [answered]
I have an XDP socket dispatcher ebpf program that declares an XSK map as follows:

```c
struct {
	__uint(type, BPF_MAP_TYPE_XSKMAP);
	__uint(key_size, sizeof(__u32));
	__uint(value_size, sizeof(__u32));
	__uint(max_entries, MAX_QUEUE_IDS);
} xsks_map SEC(".maps");
```

And an dispatcher:

```c
SEC("xdp")
int xsk_dispatcher(struct xdp_md *ctx) {
	return bpf_redirect_map(&xsks_map, ctx->rx_queue_index, XDP_PASS);
}
```

Now, in my Go code when I do this...

```go
objs := bpfObjects{}
loadBpfObjects(&objs, nil)
```

...and when doing the `loadBpfObjects` call multiple times, each time passing in a new freshly zeroed `bpfObjects{}`, do I then get multiple, independent XSK map instances?

**Comment by binw666** [comment, marked as answer]
Yes, that's right, descriptors to programs and Maps are saved in `objs`. In addition, you can redefine `bpfObjects{}` to decide which programs and Maps to load.

**Reply by ti-mo** [reply]
Also, if you need to share a map across multiple Collections (so, multiple calls to `loadBpfObjects()`) you can pin that map to bpffs to make subsequent loads use the same instance of the map.

If you prefer to keep everything in process space (no bpffs pinning), use `CollectionOptions.MapReplacements` for the *second* (re)load of your Collection after it loaded correctly once before:
```go
loadBpfObjects(&objs,
	&ebpf.CollectionOptions{
		MapReplacements: map[string]*Map{
			"xsks_map": old_objs.XsksMap,
		},
	})
```
